packages/upgrade/src/Commands/CheckPluginsCompatibilityWithV4.php

```diff
@@ -0,0 +1,208 @@
+
+        // Display result in a formatted table
+        $this->table(
+            ['Package', 'Installed Version', 'Compatible with v4?', 'Latest Compatible v4 Version'],
+            $results
+        );
+
+        return 0;
+    }
+
+    /**
+     * Checks on Packagist whether the given package has any version compatible with Filament v4.
+     *
+     * @param  string  $package  The full package name (e.g. vendor/package)
+     * @return array|null Array with version string and isPrerelease flag, or null if not found
+     */
+    protected function checkFilamentV4Compatibility(string $package): ?array
+    {
+        // First check regular releases
+        $url = "https://repo.packagist.org/p2/{$package}.json";
+        $compatibleVersion = $this->checkPackageCompatibility($package, $url);
+
+        if ($compatibleVersion !== null) {
+            return [
+                'version' => $compatibleVersion,
+                'isPrerelease' => false,
+            ];
+        }
+
+        // If no compatible version found, check dev versions
+        $devUrl = "https://repo.packagist.org/p2/{$package}~dev.json";
+        $compatibleVersion = $this->checkPackageCompatibility($package, $devUrl);
+
+        if ($compatibleVersion !== null) {
+            return [
+                'version' => $compatibleVersion,
+                'isPrerelease' => true,
+            ];
+        }
+
+        return null;
+    }
+
+    /**
+     * Checks a specific package URL for Filament v4 compatibility
+     *
+     * @param  string  $package  The full package name
+     * @param  string  $url  The Packagist API URL to check
+     * @return string|null The compatible version string or null if not found
+     */
+    protected function checkPackageCompatibility(string $package, string $url): ?string
+    {
+        try {
+            $response = Http::get($url);
+
+            if (! $response->ok()) {
+                return null;
+            }
+
+            $data = $response->json();
+            $versions = $data['packages'][$package] ?? [];
+
+            foreach ($versions as $version) {
+                $requires = $version['require'] ?? [];
+
+                if (isset($requires['filament/filament'])) {
+                    $constraint = $requires['filament/filament'];
+
+                    // Match common constraint patterns for v4
+                    if (preg_match('/\^4\.|~4\.|>=4\./', $constraint)) {
+                        return $version['version'];
+                    }
+                }
+            }
+
+        } catch (Exception $e) {
+            $this->error("Error checking {$package} at {$url}: " . $e->getMessage());
+        }
+
+        return null;
+    }
+
+    /**
+     * Checks if a package requires filament/filament
+     *
+     * @param  string  $package  The full package name
+     * @return bool True if the package requires filament/filament
+     */
+    protected function packageRequiresFilament(string $package): bool
```

Comment:
Perhaps this can be refactored to look into `vendor/{$package}/composer.json` to avoid too many requests on the Packagist server? Since it's a simple check whether the package requires Filament or not.

Reply:
great suggestion, I will do that!

Reply:
Fixed in https://github.com/filamentphp/filament/pull/17280/commits/7fbac9d8f6988bb407681c8000a1e133c91d6bbf